File: kernel/src/syscalls/sysinfo.cpp
#include "process.h"

int Process::do_sys_sysinfo(UserPtr<struct sysinfo*> info_ptr) {
	static constexpr struct sysinfo sys = {
		.uptime    = 1234,
		.loads     = {102176, 105792, 94720},
		.totalram  = 32UL*1024*1024*1024,
		.freeram   = 26UL*1024*1024*1024,
		.sharedram = 1UL *1024*1024*1024,
		.bufferram = 1UL *1024*1024*1024,
		.totalswap = 2UL *1024*1024*1024,
		.freeswap  = 2UL *1024*1024*1024,
		.procs     = 1234,
		.totalhigh = 0,
		.freehigh  = 0,
		.mem_unit  = 1
	};
	return (copy_to_user(info_ptr, &sys) ? 0 : -EFAULT);
}

File: kernel/src/process/syscalls/sysinfo.zig
usingnamespace @import("../common.zig");
const mem = @import("../../mem/mem.zig");
const UserPtr = mem.safe.UserPtr;

fn sys_sysinfo(self: *Process, info_ptr: UserPtr(*linux.sysinfo)) !void {
    const total_ram = mem.pmm.memoryLength();
    const free_ram = mem.pmm.amountFreeMemory();
    const used_ram = total_ram - free_ram;
    const total_swap = 2 * 1024 * 1024 * 1024;
    const info = linux.sysinfo{
        .uptime = 1234,
        .loads = .{ 30000, 20000, 15000 },
        .total_ram = total_ram,
        .free_ram = free_ram,
        .shared_ram = used_ram / 10,
        .buffer_ram = used_ram / 10,
        .total_swap = total_swap,
        .free_swap = total_swap,
        .procs = 1234,
        .total_high = 0,
        .free_high = 0,
        .mem_unit = 1,
    };
    try mem.safe.copyToUserSingle(linux.sysinfo, info_ptr, &info);
}

pub fn handle_sys_sysinfo(self: *Process, arg0: usize) !usize {
    const info_ptr = try UserPtr(*linux.sysinfo).fromFlat(arg0);
    try sys_sysinfo(self, info_ptr);
    return 0;
}
